### DISCUSSION TITLE
What is the correct character escaping for workflow command values e.g. echo ::error::XXXX

### DISCUSSION THREAD

**Discussion by ben-spiller** [community/community#26736] [Actions]
<p>I’m using the “echo” approach to add annotations to give errors for test failures. I’m including the error message from the test which could contain arbitrary characters, and I couldn’t see from <a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-commands-for-github-actions</a> what kind of escaping I need to perform on the command value?</p>
<p>From hacking around it looks like colons are OK but there’s something special going on with % characters (e.g. %0A produces a newline), maybe I should be doing URL encoding? But it doesn’t seem like other URL chars are encoded, so I’m not sure. Would be great if someone could definitively confirm what encoding is needed for safely representing string as workflow command values<br>
e.g.<br>
echo “::error::This is a message with chars !*’();:@&amp;=+$,/?#<span class="chcklst-box fa fa-square-o fa-fw"></span> with %25 and %0A in it”</p>

**Comment by ghost** [comment]
<p>Hello – we’re trying to reproduce this on our end:</p>
<p>Are you able to share a copy of your workflow file?</p>
<p>If the repository is public, could you link that to us as well?</p>
<p>For more context, could you let us know what the expected behaviour of the workflow run is versus what is being outputted now?</p>
<p>Thank you!</p>

**Comment by ben-spiller** [comment]
<p>Hi this isn’t a bug report so there’s nothing to repo really - I’m asking a question about how your API works so that I can make sure when I output a message using echo “::error::XXXXX" syntax the XXXXX shows up as I expect. Clearly there is some special handling for % escape sequences otherwise<br>
echo “::error::This is a message with chars !*’();:@&amp;=+$,/?# with %25 and %0A in it”<br>
would result in a message containing “%25 and %0A” rather than what it actually does which is to convert %25 to a % and %0A to a newline. I don’t think that’s a bug, but I do what to know the specification for exactly what GitHub is doing with these strings as it’s not obvious. And because I don’t want to start relying on behaviour that is not officially supported. Thanks</p>

**Comment by ben-spiller** [comment]
<p>Hi there, any input on this? Really need a definitive answer from GitHub on how this works, thanks.</p>

**Comment by ghost** [comment, marked as answer]
<p>Hi, sorry for the delay.</p>
<p>The <code>actions/toolkit</code> has a way to sanitize an input into a string – could you give these commands a try?</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/actions/toolkit/blob/cee7d92d1d02e3107c9b1387b9690b89096b67be/packages/core/src/command.ts#L92-L105" target="_blank">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/actions/toolkit/blob/cee7d92d1d02e3107c9b1387b9690b89096b67be/packages/core/src/command.ts#L92-L105" target="_blank">actions/toolkit/blob/cee7d92d1d02e3107c9b1387b9690b89096b67be/packages/core/src/command.ts#L92-L105</a></h4>
<pre class="onebox"><code class="lang-ts"><ol class="start lines" start="92" style="counter-reset: li-counter 91 ;">
<li>function escapeData(s: any): string {</li>
<li>  return toCommandValue(s)</li>
<li>    .replace(/%/g, '%25')</li>
<li>    .replace(/\r/g, '%0D')</li>
<li>    .replace(/\n/g, '%0A')</li>
<li>}</li>
<li>
</li>
<li>function escapeProperty(s: any): string {</li>
<li>  return toCommandValue(s)</li>
<li>    .replace(/%/g, '%25')</li>
<li>    .replace(/\r/g, '%0D')</li>
<li>    .replace(/\n/g, '%0A')</li>
<li>    .replace(/:/g, '%3A')</li>
<li>    .replace(/,/g, '%2C')</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by dave99galloway** [comment]
<p>Hi team - can you please give an example of how this might work in a workflow yml? I haven’t been able to figure it out from the docs</p>
<pre><code class="lang-auto"># Manually trigger a workflow to test the called workflow actions
name: Manual workflow

# Controls when the action will run. Workflow runs when manually triggered using the UI
# or API. (or push if not commented out)
on:
  workflow_dispatch:
    # Inputs the workflow accepts.
    inputs:
      reason:
        # Friendly description to be shown in the UI instead of 'reason'
        description: 'Reason for build'
        # Default value if no value is explicitly provided
        default: 'Test'
        # Input has to be provided for the workflow to run
        required: false

  jobs:
  call-workflow:
    uses: my-org/my-repo/.github/workflows/my-called-workflow.yml@MY-BRANCH
    with:
      BUILD_REASON: ${{ github.event.inputs.reason }}

</code></pre>
<p>from <a href="https://docs.github.com/en/actions/learn-github-actions/workflow-commands-for-github-actions" rel="noopener nofollow ugc">here</a> I see I need to do something like</p>
<pre><code class="lang-auto">echo '::escapeData::${{ github.event.inputs.reason }}'
</code></pre>
<p>but this gives the literal value in the called wf:-</p>
<pre><code class="lang-auto">  env:
    BUILD_REASON: echo '::escapeData::Test I've escaped this'
</code></pre>
<p>if I remove the echo, the workflow file can’t be run at all:-</p>
<pre><code class="lang-auto">      BUILD_REASON: ::escapeData::${{ github.event.inputs.reason }}

</code></pre>
<p>gives</p>
<pre><code class="lang-auto">Use workflow from
Workflow does not exist or does not have a workflow_dispatch trigger in this branch.
</code></pre>
<p>putting the command inside the deference of the env var gives:</p>
<pre><code class="lang-auto">      BUILD_REASON: ${{ ::escapeData::github.event.inputs.reason }}
 Check failure on line 39 in .github/workflows/manual.yml

GitHub Actions
/ Manual workflow
Invalid workflow file
The workflow is not valid. .github/workflows/manual.yml (Line: 39, Col: 21): Unexpected symbol: '::escapeData::github'. Located at position 1 within expression: ::escapeData::github.event.inputs.reason
</code></pre>
<p>this also fails:-</p>
<pre><code class="lang-auto">  BUILD_REASON: ${{ echo '::escape-data::github.event.inputs.reason' }}
 Check failure on line 39 in .github/workflows/manual.yml

GitHub Actions
/ Manual workflow
Invalid workflow file
The workflow is not valid. .github/workflows/manual.yml (Line: 39, Col: 21): Unrecognized named-value: 'echo'. Located at position 1 within expression: echo '::escape-data::github.event.inputs.reason'
</code></pre>

**Comment by airtower-luna** [comment]
<p>If you’re trying to pre-process the input value, you can do that in a separate job that creates an output with the escaped data. However I’d rather do any escaping that may be necessary in the called workflow, to make it more robust.</p>

**Comment by dave99galloway** [comment]
<p>hi <a class="mention" href="/airtower-luna">@airtower-luna</a> - could you please give an example ? the syntax seems weird in the examples in the doc. happy to move into the called workflow to do this, but what does that look like?</p>
<p>TIA :slight_smile:</p>

**Comment by airtower-luna** [comment]
<p>Something like this, I used <code>actions/github-script</code> to directly run Javascript code. You could do the same in any other programming language via <code>run</code>, but this way I could just copy the escape code from <code>actions/toolkit</code>. I’d have preferred to call the function, but the <a href="https://github.com/actions/toolkit/blob/cee7d92d1d02e3107c9b1387b9690b89096b67be/packages/core/src/command.ts#L3">“For internal use, subject to change” comment</a> made that look questionable. :wink:</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/airtower-luna/hello-github-actions/blob/485336f405d869bee8051f0bbfdeabedd9e4ae7f/.github/workflows/reusable-echo.yaml">
  <header class="source">

      <a href="https://github.com/airtower-luna/hello-github-actions/blob/485336f405d869bee8051f0bbfdeabedd9e4ae7f/.github/workflows/reusable-echo.yaml" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/airtower-luna/hello-github-actions/blob/485336f405d869bee8051f0bbfdeabedd9e4ae7f/.github/workflows/reusable-echo.yaml" target="_blank" rel="noopener">airtower-luna/hello-github-actions/blob/485336f405d869bee8051f0bbfdeabedd9e4ae7f/.github/workflows/reusable-echo.yaml</a></h4>


    <pre><code class="lang-yaml">name: Reusable echo example with input

on:
  workflow_call:
    inputs:
      test:
        required: true
        type: string

jobs:
  echo:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/github-script@v5.0.0
        id: pre-process
        env:
          TEST: ${{ inputs.test }}
        with:
          script: |
            const { TEST } = process.env
</code></pre>


  This file has been truncated. <a href="https://github.com/airtower-luna/hello-github-actions/blob/485336f405d869bee8051f0bbfdeabedd9e4ae7f/.github/workflows/reusable-echo.yaml" target="_blank" rel="noopener">show original</a>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by dave99galloway** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a><br>
thank you, I’m going to give this a try now. Just for my knowledge on how to use toolkit actions though, let’s say I wasn’t worried about the <a href="https://github.com/actions/toolkit/blob/cee7d92d1d02e3107c9b1387b9690b89096b67be/packages/core/src/command.ts#L3" rel="noopener nofollow ugc">“For internal use, subject to change” comment </a></p>
<p>what would the correct workflow call loo like using the built in function?</p>
<p>tia!</p>

**Comment by dave99galloway** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> - i tried your example and the echo was empty</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/dave99galloway/github-actions-issues/runs/4407583438?check_suite_focus=true">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/dave99galloway/github-actions-issues/runs/4407583438?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/60669660/181099662-41a11271-0cb3-40cc-a5e6-e811b0e4f968.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91_2_10x10.png"></div>

<h3><a href="https://github.com/dave99galloway/github-actions-issues/runs/4407583438?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">Create call-sanitiser.yml · dave99galloway/github-actions-issues@a58876b</a></h3>

  <p>reproduction of github actions issues. Contribute to dave99galloway/github-actions-issues development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>where are the step outputs supposed to be set?</p>
<p>TIA!</p>

**Comment by dave99galloway** [comment]
<p>Hi Support! do you have an example of this being used somewhere?</p>

**Comment by airtower-luna** [comment]
<p>The <code>TEST</code> environment variable is empty when the replacement is done, so of course the output will be empty, too:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/dave99galloway/github-actions-issues/runs/4407583438?check_suite_focus=true#step:2:13">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/dave99galloway/github-actions-issues/runs/4407583438?check_suite_focus=true#step:2:13" target="_blank" rel="noopener">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/5814919/181099663-0699c4bd-f960-476d-b69a-b3013dcc2693.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a20ddfa595df46b2a6d022eb1b80d849afa13c91_2_10x10.png"></div>

<h3><a href="https://github.com/dave99galloway/github-actions-issues/runs/4407583438?check_suite_focus=true#step:2:13" target="_blank" rel="noopener">Create call-sanitiser.yml · dave99galloway/github-actions-issues@a58876b</a></h3>

  <p>reproduction of github actions issues. Contribute to dave99galloway/github-actions-issues development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>The reason is here, where the parameter for the reusable workflow is set. <code>github.events</code> doesn’t exist, I assume you mean <code>github.event</code>, as <a href="https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#example-workflow-configuration">in the example</a>.</p><aside class="onebox githubactions" data-onebox-src="https://github.com/dave99galloway/github-actions-issues/actions/runs/1535065589/workflow#L21">
  <header class="source">

      <a href="https://github.com/dave99galloway/github-actions-issues/actions/runs/1535065589/workflow#L21" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
      <svg width="24" height="24" class="github-icon github-icon-success" viewbox="0 0 24 24" aria-hidden="true"><path d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12zm16.28-2.72a.75.75 0 00-1.06-1.06l-5.97 5.97-2.47-2.47a.75.75 0 00-1.06 1.06l3 3a.75.75 0 001.06 0l6.5-6.5z"></path></svg>



  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/dave99galloway/github-actions-issues/actions/runs/1535065589/workflow#L21" target="_blank" rel="noopener">Create call-sanitiser.yml</a>
    </h4>

      <div class="github-info">
        Calling Sanitiser <span class="github-run-number">#1</span>
      </div>
  </div>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
